### ISSUE TITLE
Discussion about storage index

### ISSUE THREAD

**Issue by weijietong** [apache/incubator-druid#4641]
Druid is not good at querying the dimensions data out as our experience. This limits some usage scenarios which depend on the detail dimension column not the metric column like join query.Here I fired this try to discuss a solution to fix this bad performance. 

I think the storage format should have some improvements to support this. When we query the dimension column value,we go through two steps: 1st we go through the forward index to find the dictionary id according to the row number getting from the corresponding bitmap,then we find the real dictionary value by the dictionary id through the GenericIndex. If the dimension column is high cardinality,it will exhaust much CPU periods.The above two steps to fetch a value is not CPU pipeline ,multi level cache friendly . I ever do a comparison between the Druid index and a b plus tree implementation to fetch the dimension data.says 100w row number,40w cardinality a column per segment,I queried out 10w dimension data .The result shows that Druid needs 50ms ,btree needs 2ms(query from row number 1 to 10w ). As btree stores row number as key,the dimension value as leaf node,when It find the corresponding leaf nodes,it just needs to iterate through the linked leaf nodes.so that behavior is high performance.

I wonder whether Druid could add another storage block to store dimension data additionally by still keeping the dictionary to filter bit maps. Or is there any other ideas ?

**Comment by gianm** [comment]
Interesting question.

What did the btree implementation look like? Was it hooked into Druid at all or did you do it as a separate thing? Was it reading from mmapped something into java.lang.Strings, like Druid would?

What I remember from looking at profiles is that when reading dimension data, a big chunk of the time is spent decoding UTF8 into java.lang.String. It could be cached but is not by default (see `druid.processing.columnCache.sizeBytes`).

I was also considering that for some cases (like the Scan or Select queries) we could improve throughput by doing these operations in chunks. So we could read 1000 rows from a dimension column and then go and lookup all the names later. CPU cache behavior should be better, and also we may save some UTF-8 decodes if the 1000 rows have fewer than 1000 unique values. This can work even if the column-cache from `druid.processing.columnCache.sizeBytes` is disabled.

**Comment by gianm** [comment]
I think adding another (optional) section to Druid string columns is good if it's going to give a benefit. It should be optional so people that don't want it, don't need to store it.

Also wondering what you think about the chunk operation suggestion.

**Comment by weijietong** [comment]
@gianm thanks for response.

 Actually I have done the profile and tuned the codes with the methods what you mentioned before. Yes, the profile shows that string decode and GenericIndex.get is the hot bottleneck. I ever fetched the dictionary string bytes directly out from the mapped ByteBuffer without decoding.This will send the pure bytes to the front end console.But no noticeable performance gains. Chunk operations also been tried. I first cached a batch of (row number --> dictionary id）pairs,then batch looked up the dictionary value. The performance gains was also not noticeable.

I think the reason is that batched dictionary ids may cross different cpu cache blocks. To a high cardinality dimension column , looking up the dictionary value is a randomized seeking operation not sequence reading(I think sequence reading is the key point why metric columns perform so good , it will leverage the CPU prefetched data,no cache switches ). The parquet project also adopts the dictionary encoding.But it recommends cutting the rows numbers into separate row groups,each row group's column data size is CPU cache friendly by using dictionary encoding. But this method may not apply to Druid,as Druid needs a global dictionary to a segment to do the filter operation(Of cause,maybe we can have some other tricks to solve it .welcome to discuss that).

I do the exercises out of the Druid by using the jdbm3 project. One column a btree, using the mmap too . Once we got the filtered bitmap(meaning selected row numbers),we just need to browse the matched btree leaf nodes(notice that browse method,not get) . The underneath operation is sequence reading linked dimension value. If we add this additional storage,option strategy is indeed , as user should pay for the additional disk storage space.
